packages/flutter/lib/src/rendering/object.dart

```diff
@@ -378,28 +389,51 @@ class PaintingContext extends ClipContext {
 
   /// Clip further painting using a rectangle.
   ///
+  /// {@template flutter.rendering.object.needsCompositing}
   /// * `needsCompositing` is whether the child needs compositing. Typically
-  ///   matches the value of [RenderObject.needsCompositing] for the caller.
+  ///   matches the value of [RenderObject.needsCompositing] for the caller. If
+  ///   false, this method returns null, indicating that a layer is no longer
+  ///   necessary. If a render object calling this method stores the `oldLayer`
```

Comment:
Should a render object always set its [RenderObject.layer] field to the result of this method to no matter if this method returns null or not?

Reply:
There's a many-to-many relationship between render objects and layers. Requiring setting it to whatever a `push*` method returns implies a one-to-one relationship. Instead, setting `RenderObject.layer` is optional. You set it to get some performance benefits, but that's it. There needs to be a logical "root" layer that's stable across frames for `RenderObject.layer` to make sense.